src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,658 @@
+            }
+            else
+            {
+                _started = StartState.None;
+
+                _canWrite = true;
+                _canRead = !flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL);
+            }
+
+            // Options to effectively have a single buffer pipe
+            // Effectively this is a synchronization mechanism between calls to ReadAsync and Recv callbacks
+            // However, having a pipe here is nice for calling Complete,
+            // handing buffering, and handling synchronization.
+            _sendResettableCompletionSource = new UIntResettableCompletionSource();
+            _receiveResettableCompletionSource = new UIntResettableCompletionSource();
+            SetCallbackHandler();
+        }
+
+        internal override bool CanRead => _canRead;
+
+        internal override bool CanWrite => _canWrite;
+
+        internal override long StreamId
+        {
+            get
+            {
+                if (_streamId == -1)
+                {
+                    _streamId = GetStreamId();
+                }
+
+                return _streamId;
+            }
+        }
+
+        internal override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (_started == StartState.None)
+            {
+                _started = StartState.Started;
+                await StartAsync();
+            }
+
+            await SendAsync(new ReadOnlySequence<byte>(buffer), QUIC_SEND_FLAG.NONE);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+        }
+
+        internal override async ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            // TODO is there anything we can do with cancellation token?
+            lock (_sync)
+            {
+                if (_readState == ReadState.ReadsCompleted)
+                {
+                    if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+                    return 0;
+                }
+                else if (_readState == ReadState.ReadsAborted)
+                {
+                    throw new IOException("Reading has been aborted by the peer.");
+                }
+            }
+
+            // TODO there could potentially be a perf gain by storing the buffer from the inital read
+            // This reduces the amount of async calls, however it makes it so MsQuic holds onto the buffers
+            // longer than it needs to. We will need to benchmark this.
+            int length = (int)await _receiveResettableCompletionSource.GetValueTask();
+            if (NetEventSource.IsEnabled) NetEventSource.Info("Read completed");
+
+            static unsafe void CopyToBuffer(Span<byte> buffer, QuicBuffer[] quicBuffers)
```

Comment:
I'd suggest either inlining this code, or to name the parameters `destinationBuffer` and `sourceBuffers` for better clarity.